src/defs.hpp

```diff
@@ -56,21 +60,211 @@ struct LogicalLocation { // aggregate and POD type
+    std::vector<LogicalLocation> locs;
+
+    const auto irange =
+        (std::abs(ox1) == 1) ? std::vector<int>{0, ox1} : std::vector<int>{0};
+    const auto jrange =
+        (std::abs(ox2) == 1) ? std::vector<int>{0, ox2} : std::vector<int>{0};
+    const auto krange =
+        (std::abs(ox3) == 1) ? std::vector<int>{0, ox3} : std::vector<int>{0};
+
+    auto AddNeighbors = [&](const LogicalLocation &loc) {
+      int n_cells_level = std::pow(2, loc.level());
+      for (int i : irange) {
+        for (int j : jrange) {
+          for (int k : krange) {
+            const auto lx1 = loc.lx1() + i;
+            const auto lx2 = loc.lx2() + j;
+            const auto lx3 = loc.lx3() + k;
+            // TODO(LFR): Deal with periodic boundaries, maybe a little complicated
+            // because of root grid stuff
+            if (0 <= lx1 && lx1 < n_cells_level && 0 <= lx2 && lx2 < n_cells_level &&
+                0 <= lx3 && lx3 < n_cells_level) {
+              locs.emplace_back(loc.level(), lx1, lx2, lx3);
+              auto parent = locs.back().GetParent();
+              if (IsNeighbor(parent)) locs.push_back(parent);
+            }
+          }
+        }
+      }
+    };
+
+    AddNeighbors(*this);
+
+    // Iterate over daughters of this block that share the same topological element
+    for (int l :
+         (std::abs(ox1) == 1) ? std::vector<int>{ox1 > 0} : std::vector<int>{0, 1}) {
+      for (int m :
+           (std::abs(ox2) == 1) ? std::vector<int>{ox2 > 0} : std::vector<int>{0, 1}) {
+        for (int n :
+             (std::abs(ox3) == 1) ? std::vector<int>{ox3 > 0} : std::vector<int>{0, 1}) {
+          AddNeighbors(GetDaughter(l, m, n));
+        }
+      }
+    }
+    // The above procedure likely duplicated some blocks, so put them in a set
+    return std::set<LogicalLocation>(std::begin(locs), std::end(locs));
   }
 };
 
+inline bool operator<(const LogicalLocation &lhs, const LogicalLocation &rhs) {
+  if (lhs.morton() == rhs.morton()) return lhs.level() < rhs.level();
+  return lhs.morton() < rhs.morton();
+}
+
+inline bool operator>(const LogicalLocation &lhs, const LogicalLocation &rhs) {
+  if (lhs.morton() == rhs.morton()) return lhs.level() > rhs.level();
+  return lhs.morton() > rhs.morton();
+}
+
+inline bool operator==(const LogicalLocation &lhs, const LogicalLocation &rhs) {
+  return ((lhs.level() == rhs.level()) && (lhs.lx1() == rhs.lx1()) &&
+          (lhs.lx2() == rhs.lx2()) && (lhs.lx3() == rhs.lx3()));
+}
+
+struct block_ownership_t {
+ public:
+  const bool &operator()(int ox1, int ox2, int ox3) const {
+    return ownership[ox1 + 1][ox2 + 1][ox3 + 1];
+  }
+  bool &operator()(int ox1, int ox2, int ox3) {
+    return ownership[ox1 + 1][ox2 + 1][ox3 + 1];
+  }
+
+ private:
+  bool ownership[3][3][3];
+};
+
+inline block_ownership_t
+DetermineOwnership(const LogicalLocation &main_block,
+                   const std::set<LogicalLocation> &allowed_neighbors) {
+  block_ownership_t main_owns;
```

Comment:
the above ownership stuff can probably also go in the above-mentioned header

Reply:
done